api/model/src/main/java/org/projectnessie/api/v2/http/HttpTreeApi.java

```diff
@@ -855,8 +855,8 @@ MergeResponse mergeRefIntoBranch(
   @Override
   @POST
   @jakarta.ws.rs.POST
-  @Path("{branch}/history/commit")
-  @jakarta.ws.rs.Path("{branch}/history/commit")
+  @Path("{branch:" + REF_NAME_PATH_ELEMENT_REGEX + "}/history/commit")
+  @jakarta.ws.rs.Path("{branch:" + REF_NAME_PATH_ELEMENT_REGEX + "}/history/commit")
```

Comment:
Would it make sense to do similar changes for merge and transplant endpoints?

Reply:
For merge & transplant a relative hash doesn't really make any sense,

The old data model was relatively strict wrt `expectedHash` - unnecessarily actually.

It feels, that the merge-operation, as it exists today using a squashed commit based on the diff between the merge-base and the merge-source, does not need any expected-hash. The "expected state" and the "applied change" are implicit (present in the Nessie repo), not explicit (only the new value provided by the caller). Older implementations, that supported "merges w/ individual commits", the expected-hash was important, because it was effectively a transplant.

The implementation for the new data model doesn't use the expected-hash for merge nor for transplant. For the latter (transplant) that seems wrong though. @adutra do you have bandwidth to check whether the expected-hash is required for transplant (I think it is, but I might be wrong).

In any case, every committing operation that takes an expected-hash, should not allow relative commit IDs. Good catch!

Reply:
From what I see, in the new model:

* **For all committing operations (commit, merge, transplant)**: if an expected-hash is provided, an expected `CommitObj` is loaded, if the object does not exist, an exception is thrown:

https://github.com/projectnessie/nessie/blob/795ba1da4cf630d40dca8b967c79c6ae2daa40eb/versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/BaseCommitHelper.java#L136-L142

* Other than that, the expected-hash itself is only used for reporting purposes (when creating error messages, and `MergeResult` instances).

* **For commits only**: the expected `CommitObj` is further used to load the so-called expected-index:

https://github.com/projectnessie/nessie/blob/c6388c3fd3120a23c8db1516369d1ebb478599da/versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/CommitImpl.java#L117-L124

This index is then used to compute key conflicts.

* **For merges and transplants**: I do not see any further usage of expected-hash or the corresponding expected `CommitObj`.

Reply:
> The implementation for the new data model doesn't use the expected-hash for merge nor for transplant. For the latter (transplant) that seems wrong though.

Why is it not wrong for merges? If you are merging into some branch, expecting its head to be a certain hash, shouldn't the system raise an error if the head of the branch has moved away?

Reply:
With "not wrong" I meant that a diff-based squash-merge does not need it to work correctly.
Users _may_ want to only apply a squash, if the HEAD is exactly at that hash.

Reply:
https://github.com/projectnessie/nessie/issues/7202